### PR TITLE
Fix ComputationalMetric.forward() to not only consider the last batch

### LINKED ISSUE #644: CompositionalMetric only returns value for last batch
## 🐛 Bug

When using a compositional metric, i.e., `metric_a + metric_b` or `2 * metric_a` with `forward()` instead of `update()`, the value returned by `.compute()` will be based only on the last batch.

This is because the default implementation of `Metric.forward()`, which is inherited by `CompositionalMetric`, calls `reset()`, which in turn resets all child metrics. Their state, however, is not restored by `forward()` afterwards.

My proposed fix would be to override `forward()` in `CompositionalMetric` and I will file a PR implementing this.

#### Code sample

```python
import torch
import torchmetrics
from torchmetrics.metric import CompositionalMetric

# Example predictions and targets
pred = torch.tensor([0.4, 0.2, 0.6, 0.7, 0.5])
target = torch.tensor([0.5, 0.4, 0.4, 0.3, 0.6])

# Compute base metric in batches
metric = torchmetrics.MeanSquaredError()
metric(preds=pred[:3], target=target[:3])
metric(preds=pred[3:], target=target[3:])
val = metric.compute()
print(val)

# Compute compositional metric in batches
cm = 1 * torchmetrics.MeanSquaredError()
cm(preds=pred[:3], target=target[:3])
cm(preds=pred[3:], target=target[3:])
val = cm.compute()
print(val)
```

### Expected behavior

Two identical values are printed:

```
tensor(0.0520)
tensor(0.0520)
```

### Actual (faulty) behavior

The compositional metric only returns the value for the last batch:

```
tensor(0.0520)
tensor(0.0850)
```

### Environment

- PyTorch Version (e.g., 1.0): 1.10.0
- OS (e.g., Linux): Linux
- How you installed PyTorch (`conda`, `pip`, source): conda
- Build command you used (if compiling from source):
- Python version: 3.8
- CUDA/cuDNN version: 11.1
- GPU models and configuration:
- Any other relevant information:

### PR DESCRIPTION
**Pull request by Callidior**
## What does this PR do?

Fixes #644 

## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: tests/bases/test_composition.py [hunk 1]
```diff
@@ -557,3 +553,19 @@ def test_compositional_metrics_update():
...
+
+
+def test_compositional_metrics_forward():
+
+    compos = DummyMetric(5) + DummyMetric(4)
+
+    assert isinstance(compos, CompositionalMetric)
+    compos()
+    compos()
+    compos()
```
Comment (SkafteNicki) [inline tests/bases/test_composition.py:565]: lets also check that the return value is correct
```suggestion
    assert compos() == 9
    assert compos() == 9
    assert compos() == 9
```

File: tests/bases/test_composition.py [hunk 2]
```diff
@@ -557,3 +553,19 @@ def test_compositional_metrics_update():
 
     assert compos.metric_a._num_updates == 3
     assert compos.metric_b._num_updates == 3
+
+
+def test_compositional_metrics_forward():
+
+    compos = DummyMetric(5) + DummyMetric(4)
```
Comment (SkafteNicki) [inline tests/bases/test_composition.py:560]: lets try to test composition with metric and something else
```suggestion
@pytest.mark.parametrize("metric_b", [4, DummyMetric(4)])
def test_compositional_metrics_forward(metric_b):
    """ test forward method of compositional metric """
    compos = DummyMetric(5) + metric_b
```

**Comment by Callidior** [comment]
@SkafteNicki Thanks for your suggestions! I committed them as proposed.

**Comment by SkafteNicki** [comment]
Hi @Callidior,
Just add changelog and improved one of test to get a bit better coverage :]

**Review by SkafteNicki — APPROVED** [review]

File: torchmetrics/metric.py [hunk 3]
```diff
@@ -756,6 +756,34 @@ def compute(self) -> Any:
 
         return self.op(val_a, val_b)
 
+    @torch.jit.unused
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+
+        if isinstance(self.metric_a, Metric):
+            val_a = self.metric_a(*args, **self.metric_a._filter_kwargs(**kwargs))
+        else:
+            val_a = self.metric_a
```
Comment (Borda) [inline torchmetrics/metric.py:765]:
```suggestion
        val_a = self.metric_a(*args, **self.metric_a._filter_kwargs(**kwargs)) if isinstance(self.metric_a, Metric) else self.metric_a
```

File: torchmetrics/metric.py [hunk 4]
```diff
@@ -756,6 +756,34 @@ def compute(self) -> Any:
...
+
+        if isinstance(self.metric_b, Metric):
+            val_b = self.metric_b(*args, **self.metric_b._filter_kwargs(**kwargs))
+        else:
+            val_b = self.metric_b
```
Comment (Borda) [inline torchmetrics/metric.py:770]:
```suggestion
        val_b = self.metric_b(*args, **self.metric_b._filter_kwargs(**kwargs)) if isinstance(self.metric_b, Metric) else self.metric_b
```

File: torchmetrics/metric.py [hunk 5]
```diff
@@ -756,6 +756,36 @@ def compute(self) -> Any:
...
 
+    @torch.jit.unused
+    def forward(self, *args: Any, **kwargs: Any) -> Any:
+
+        val_a = (
+            self.metric_a(*args, **self.metric_a._filter_kwargs(**kwargs))
+            if isinstance(self.metric_a, Metric)
+            else self.metric_a
+        )
+
```
Comment (Borda) [inline torchmetrics/metric.py:767]:
```suggestion
```

**Review by Borda — APPROVED** [review]
